### PR TITLE
[Core][LockObject] Added omp_test_lock interface to LockObject

### PR DESCRIPTION
**Pull request by emiroglu**
**Description**
This PR implements the interface to `omp_test_lock` interface to `LockObject` class.

The intent is to use in finite differencing schemes, where the nodal variables are perturbed. Currently, the only way of implementation is through defining a critical region, which slows down finite differencing with big models, or locking element nodes, which can cause a deadlock.

Using `TestLock` it is possible to try locking the element nodes and when the lock is not successful unlock all the locked nodes, then try locking them again, since it does not block the code execution.

**Changelog**
- Added an interface to `omp_test_lock` to `LockObject`

### CODE REVIEW

**Comment by philbucher** [comment]
@emiroglu did you try this if it actually works then in your implementation?

**Comment by RiccardoRossi** [comment]
As an aside, this is not an atomic operation. I am not sure of what would be the correspondent in c++11

**Comment by philbucher** [comment]
> As an aside, this is not an atomic operation. I am not sure of what would be the correspondent in c++11

[there is](https://en.cppreference.com/w/cpp/thread/mutex/try_lock)

**Comment by RiccardoRossi** [comment]
ok then, but ... at this point shouldn't we return a reference to the lock_object itself?

**Comment by RiccardoRossi** [comment]
also...are we going to use a mutex? 

because each mutex is 24 bytes...that is 3 doubles

https://stackoverflow.com/questions/16693992/why-is-sizeofstdmutex-40-gcc-clang-icc

we have the option of using a spinlock which would be 1 double ... at least we need to consider it

**Comment by philbucher** [comment]
> also...are we going to use a mutex?

I don't know
=> waiting for feedback in #8161

**Comment by emiroglu** [comment]
See the last commit in #8174 which removes the deadlock problem in combination with the `omp_test_lock`

**Comment by philbucher** [comment]
@RiccardoRossi [here](https://rigtorp.se/spinlock/) is an implementation of a spinlock with `try_wait`

**Comment by RiccardoRossi** [comment]
@philbucher this looks very appealing, however it uses a X86 only operation so we will need to be a bit careful about that.

the bigger advantage i see over a mutex is that the mutex is a heavier object, so aside for the performance overhead it will also imply making the node heavier in terms of memory occupation

**Comment by philbucher** [comment]
> @philbucher this looks very appealing, however it uses a X86 only operation so we will need to be a bit careful about that.
> 
> the bigger advantage i see over a mutex is that the mutex is a heavier object, so aside for the performance overhead it will also imply making the node heavier in terms of memory occupation

This is pretty much the first link I found on google :D

I basically wanted to see if the try_lock is possible with a spinlock

File: kratos/includes/lock_object.h [hunk 1]
```diff
@@ -100,6 +100,15 @@ namespace Kratos
...
 		}
 
+    inline int TestLock() const
+    {
+      // returns successful lock if openMP is not present
+#ifdef KRATOS_SMP_OPENMP
+      return omp_test_lock(&mLock);
+#endif
+      return 1;
+    }
```
Comment (philbucher) [inline kratos/includes/lock_object.h]:
```suggestion
    inline bool TestLock() const
    {
#ifdef KRATOS_SMP_OPENMP
      return omp_test_lock(&mLock);
#endif
      return true;
    }
```
more compliant with the upcoming changes for C++ based parallelism

**Review by philbucher — COMMENTED** [review]
@RiccardoRossi @KratosMultiphysics/technical-committee I did some research (also see #8161) and this functionality will be also available with C++11 based parallelism without much struggle (both for `std::mutex` and `spinlock`)

Hence from my side this PR is a go, as it solves a long standing problem with parallelizing loops over Elements/Conditions where all of the nodes need to be locked such as the usecase @emiroglu mentioned

FYI @armingeiser @MFusseder this will also allow you to parallelize the sensitivity computation for shape sensitivities

**Comment by emiroglu** [comment]
It would actually make sense to add LockNodes and UnLockNodes functions to the element base class so that we avoid multiple implementations.

[Here](https://github.com/KratosMultiphysics/Kratos/blob/90c385c74e8ab56ccacce6bd2eccd36c9f819575/applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp#L408-L458) is the implementation btw in case it is interesting @armingeiser @MFusseder

**Comment by philbucher** [comment]
Update:
This PR is one step to make the `LockObject` compliant with [`Lockable`](https://en.cppreference.com/w/cpp/named_req/Lockable)
Which is pretty cool I think, aligns with the other C++ locks

**Comment by emiroglu** [comment]
@KratosMultiphysics/technical-committee any final comments on this feature?

**Review by philbucher — APPROVED** [review]
@KratosMultiphysics/technical-committee approves this, it fits well with our future plans
Thanks @emiroglu

**Pull request merged**